File: kotlin/kotlin-vs-java/java-example/src/main/java/org/example/ex05/Exam05.java
package org.example.ex05;

import java.util.List;
import java.util.function.Predicate;

/**
 * 고차함수
 */
public class Exam05 {

    private Predicate<String> stringPredicate = new Predicate<String>() {
        @Override
        public boolean test(String s) {
            return s.equals("a");
        }
    };

    public Exam05() {

        var strList = List.of("a", "b", "c", "d", "e");

        var result = strList.stream()
                .filter(it -> it.equals("a"))
                .toList();
    }

    public static void main(String[] args) {
        new Exam05();
    }
}

File: kotlin/kotlin-vs-java/kotlin-example/src/main/kotlin/ex05/Exam05.kt
package ex05

import java.util.function.Predicate

/**
 * 고차함수
 */
fun main() {

    val numberList = listOf(1, 2, 3, 4, 5)

//    val pred = object : Predicate<Int> {
//
//        override fun test(t: Int): Boolean {
//            return t % 2 == 0
//        }
//
//    }

    val pred = Predicate<Int> { t -> t % 2 == 0 }

    numberList.filter { it % 2 == 0 }
    numberList.stream().filter(pred)

    val add = { x: Int, y: Int -> x + y }

    println(add(2, 3))
    println(add.invoke(2, 3))

    val _add = fun(x: Int, y: Int): Int {
        return x + y
    }

    println(_add(3, 4))

    lambda(4, 5, _add)
}

fun lambda(x: Int, y: Int, method: (Int, Int) -> Int) {
    println(method(x, y))
}